feat: remove dynamic schema generation and `search()` function

neilyio opened the pull request:
## What

This PR removes the dynamic schema generation previously used in `pg_search` and fully adopts the `@@@` operator model. It simplifies the indexing and querying process by eliminating the need for dynamically generated schemas and functions for each index.

Key changes include:
- Removal of per-index schemas and associated functions
- Modification of core functions to accept index names as parameters
- Updates to tests to reflect the new approach
- Upgrade script to transition existing indexes to the new model

## Why

The previous dynamic schema generation approach was more complicated than necessary. It created a large number of schemas and functions, cluttering the database. `@@@`, enabled by @eeeebbbbrrrr 's custom scan work, is much more concise and performant.

## How

The implementation involved several key steps:
1. Removing code related to dynamic schema generation
2. Modifying core functions like `schema`, `index_size`, and others to accept index names as parameters
3. Updating the `create_bm25` function to no longer generate schemas and functions
4. Adjusting all tests to work with the new approach
5. Creating an upgrade script (`pg_search--0.10.3--0.10.4.sql`) to transition existing indexes

The upgrade script is particularly important as it:
- Drops old functions that are no longer needed
- Creates new versions of core functions that work with the @@@ model
- Ensures backward compatibility for existing indexes

## Tests

All existing tests have been updated to work with the new model. The changes include:
- Modifying test queries to use the @@@ operator instead of dynamically generated functions
- Updating assertions to match the new query results
- Removing tests that were specific to the old dynamic schema approach
- Adding new tests to ensure the core functionality works as expected with the new model

Additionally, manual testing has been performed to ensure:
- Existing indexes continue to work after the upgrade
- New indexes can be created and used effectively
- Performance is maintained or improved

eeeebbbbrrrr commented on pg_search/src/api/index.rs:
```diff
@@ -34,7 +34,7 @@ use tantivy::schema::{FieldType, OwnedValue, Value};
 #[pg_extern]
-pub fn schema_bm25(
+pub fn schema(
     index_name: &str,
```
I would prefer this function take a `PgRelation` instead of a string.  The reason is that Postgres will validate the user is providing a valid, and fully-qualified (or SEARCH_PATH included) relation name for us.

```suggestion
    index_name: PgRelation,
```

However, we'll also need to reopen the relation at the start of the function with an AccessShareLock:

```rust
let index_name = PgRelation::with_lock(index_name.oid(), pg_sys::AccessShareLock as _);
```

The reason for reopening it is to ensure a concurrent DROP/ALTER doesn't come in while we're reading it.

eeeebbbbrrrr commented on pg_search/src/bootstrap/create_bm25.rs:
```diff
@@ -465,14 +275,9 @@ unsafe fn delete_bm25_index_by_oid(index_oid: pg_sys::Oid) -> Result<()> {
+#[pg_extern]
 fn index_size(index_name: &str) -> Result<i64> {
-    let index_oid = index_oid_from_index_name(&format!("{index_name}_bm25_index"));
+    let index_oid = index_oid_from_index_name(index_name);
```
I would prefer this function take a `PgRelation` instead of a string.  The reason is that Postgres will validate the user is providing a valid, and fully-qualified (or SEARCH_PATH included) relation name for us.

```suggestion
    index_name: PgRelation,
```

However, we'll also need to reopen the relation at the start of the function with an AccessShareLock:

```rust
let index_name = PgRelation::with_lock(index_name.oid(), pg_sys::AccessShareLock as _);
```

The reason for reopening it is to ensure a concurrent DROP/ALTER doesn't come in while we're reading it.

eeeebbbbrrrr reviewed (commented):
@neilyio a few minor things to adjust.

Also, I'm not seeing in the upgrade script where the `.search()` and `.score_bm25()` functions are being dropped?

neilyio commented:
@eeeebbbbrrrr thanks for the code examples and your `PgRelation` suggestion. I've implemented those and added some tests for the error cases.

As for the missing `DROP .search()/score_bm25()`, how do you think that should look in the migration SQL? Those are attached to the dynamically generated schemas (not `paradedb`).

eeeebbbbrrrr commented:
> As for the missing DROP .search()/score_bm25(), how do you think that should look in the migration SQL? Those are attached to the dynamically generated schemas (not paradedb).

oh. Of course. Duh. Hmm

yeah, I guess there’s nothing for us to do. 

I’m sure we could write some plpgsql function to find them and drop them, but we can let users do that, I guess. Feels a little risky for us to write code that goes and drops user schema objects.

eeeebbbbrrrr commented:
@neilyio inthink the upgrade script needs updated to reflect the argument type changes to regclass. 

I’m reviewing on my phone so if I’m reading the diff wrong, please ignore me.

eeeebbbbrrrr reviewed (approved):
lets ship it!
